tao_compiler/mlir/disc/transforms/disc_compare_simplifier.cc

```diff
@@ -0,0 +1,363 @@
+             ", isKnownNotNegative: " + bool2str(isKnownNotNegative()) +
+             ", isKnownNotPositive: " + bool2str(isKnownNotPositive());
+    }
+  };
+
+ public:
+  // TODO: plan to use our designed data structure (i.e., ShapeAnalysisV2) in
+  // the near future.
+  using ShapeOrValueInfo = ShapeComponentAnalysis::ShapeOrValueInfo;
+  using Symbol = ShapeComponentAnalysis::Symbol;
+  using SymbolicExpr = ShapeComponentAnalysis::SymbolicExpr;
+  using SymbolicExprsMap = ShapeComponentAnalysis::SymbolicExprsMap;
+
+ private:
+  // The following functions about AffineExpr checking will be packaged in
+  // `ShapeAnalysisV2` in the future.
+  SignState getSignState(const Symbol& symbol) const;
+  SignState getSignState(const AffineExpr& expr,
+                         const SmallVectorImpl<Symbol>& symbols) const;
+  bool isSpecificConstant(const SymbolicExpr& symbolExpr, int64_t value) const;
+
+  llvm::Optional<bool> compareSymbolicExpr(
+      const SymbolicExpr& lhs, const SymbolicExpr& rhs,
+      arith::CmpIPredicate predicate) const;
+};
+
+ArithComISimplifier::SignState ArithComISimplifier::getSignState(
+    const ArithComISimplifier::Symbol& symbol) const {
+  SignState state;
+  int64_t index = symbol.index;
+  // If the symbol is coming from a shape, it can't be negative. Also allow
+  // results of shape_of, compute_reshape_shape, num_elements and broadcast.
+  // This is correct, not complete.
+  if (symbol.source.isShapeInfo()) {
+    auto type = symbol.source.value().getType().dyn_cast<RankedTensorType>();
+    if (type && type.getDimSize(index) > 0) {
+      state.knownPositive = true;
+    } else {
+      state.knownNotNegative = true;
+    }
+  }
+  Operation* op = symbol.source.value().getDefiningOp();
+  if (op == nullptr) {
+    return state;
+  }
+  if (llvm::isa<shape::ShapeOfOp, mhlo::ComputeReshapeShapeOp,
+                shape::NumElementsOp, shape::BroadcastOp>(op)) {
+    // TODO: double-check shape.broadcast. If the semantics allow negative
+    // value, check whether the broadcast is a producer of mhlo.bcast/reshape
+    // op.
+    state.knownNotNegative = true;
+  } else if (auto select = dyn_cast_or_null<arith::SelectOp>(op)) {
+    auto trueValue = select.getTrueValue();
+    auto falseValue = select.getFalseValue();
+    ShapeComponentAnalysis shapeComponentAnalysis;
+    auto trueInfo = shapeComponentAnalysis.GetValueInfo(trueValue);
+    auto falseInfo = shapeComponentAnalysis.GetValueInfo(falseValue);
+    if (trueInfo && ((*trueInfo).size() == 1) && falseInfo &&
+        ((*falseInfo).size() == 1)) {
+      auto trueSymbolicExpr = (*trueInfo)[0];
+      auto falseSymbolicExpr = (*falseInfo)[0];
+      auto trueState =
+          getSignState(trueSymbolicExpr.expr, trueSymbolicExpr.symbols);
+      auto falseState =
+          getSignState(falseSymbolicExpr.expr, falseSymbolicExpr.symbols);
+      if (trueState.isKnownZero() && falseState.isKnownZero()) {
```

Comment:
```
if (trueState == falseState) {
  state = trueState;
}
```

is this enough for short?

Reply:
We still need the separated ones when trueState is not the same with falseState.